api/src/test/java/org/pmiops/workbench/billing/FreeTierBillingServiceTest.java

```diff
@@ -55,221 +59,617 @@
+
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verify(notificationService).alertUserFreeTierExpiration(eq(user));
+    assertSingleWorkspaceTestDbState(user, workspace, BillingStatus.INACTIVE, 100.01);
+
+    // expire due to time
+
+    workbenchConfig.billing.defaultFreeCreditsDaysLimit = 10;
+    final Instant tooLongAgo = Instant.now().minus(Period.ofDays(11));
+    user.setFirstRegistrationCompletionTime(Timestamp.from(tooLongAgo));
+    userDao.save(user);
+
+    // check that we do not alert twice
+
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verifyZeroInteractions(notificationService);
+  }
+
+  @Test
+  public void checkFreeTierBillingUsage_exceedsBothLimitsTimeFirst() {
+    DbUser user = createUser(SINGLE_WORKSPACE_TEST_USER);
+    DbWorkspace workspace = createWorkspace(user, SINGLE_WORKSPACE_TEST_PROJECT);
+    workbenchConfig.billing.defaultFreeCreditsDollarLimit = 100.0;
+    doReturn(mockBQTableSingleResult(0.0)).when(bigQueryService).executeQuery(any());
+
+    // expire due to time
+
+    workbenchConfig.billing.defaultFreeCreditsDaysLimit = 10;
+    final Instant tooLongAgo = Instant.now().minus(Period.ofDays(11));
+    user.setFirstRegistrationCompletionTime(Timestamp.from(tooLongAgo));
+    userDao.save(user);
+
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verify(notificationService).alertUserFreeTierExpiration(eq(user));
+    assertSingleWorkspaceTestDbState(user, workspace, BillingStatus.INACTIVE, 0.0);
+
+    // expire due to cost
 
-    doReturn(tableResult).when(bigQueryService).executeQuery(any());
+    doReturn(mockBQTableSingleResult(100.01)).when(bigQueryService).executeQuery(any());
+
+    // check that we do not alert twice
+
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verifyZeroInteractions(notificationService);
   }
 
   @Test
-  public void checkFreeTierBillingUsage_singleProjectExceedsLimit() {
+  public void checkFreeTierBillingUsage_timeExpirationAndCostThreshold() {
+
+    // test the behavior when the user exceeds the free credit time limit
+    // and also crosses the 50% cost threshold
+
+    DbUser user = createUser(SINGLE_WORKSPACE_TEST_USER);
+    DbWorkspace workspace = createWorkspace(user, SINGLE_WORKSPACE_TEST_PROJECT);
     workbenchConfig.billing.defaultFreeCreditsDollarLimit = 100.0;
+    doReturn(mockBQTableSingleResult(50.1)).when(bigQueryService).executeQuery(any());
+
+    workbenchConfig.billing.defaultFreeCreditsDaysLimit = 10;
+    final Instant tooLongAgo = Instant.now().minus(Period.ofDays(11));
+    user.setFirstRegistrationCompletionTime(Timestamp.from(tooLongAgo));
+    userDao.save(user);
+
+    // we expect to see ONE alert due to time expiration
+    // and NO alert for crossing the 50% cost threshold
+
+    freeTierBillingService.checkFreeTierBillingUsage();
+    verify(notificationService, times(1)).alertUserFreeTierExpiration(eq(user));
+    verify(notificationService, times(0))
```

Comment:
nit - the args for this mock can be any()

Reply:
ah nice